File: zenscript-language-server/src/main/java/raylras/zen/lsp/ZenLanguageService.java
package raylras.zen.lsp;

import org.eclipse.lsp4j.*;
import org.eclipse.lsp4j.jsonrpc.messages.Either;
import org.eclipse.lsp4j.services.TextDocumentService;
import org.eclipse.lsp4j.services.WorkspaceService;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;
import raylras.zen.lsp.provider.*;
import raylras.zen.model.CompilationEnvironment;
import raylras.zen.model.CompilationUnit;
import raylras.zen.model.Compilations;
import raylras.zen.util.LogMessages;
import raylras.zen.util.PathUtil;
import raylras.zen.util.l10n.L10N;

import java.nio.file.Files;
import java.nio.file.Path;
import java.util.HashSet;
import java.util.List;
import java.util.Optional;
import java.util.Set;
import java.util.concurrent.CompletableFuture;
import java.util.concurrent.locks.ReadWriteLock;
import java.util.concurrent.locks.ReentrantReadWriteLock;

public class ZenLanguageService implements TextDocumentService, WorkspaceService {

    private static final Logger logger = LoggerFactory.getLogger(ZenLanguageService.class);

    private final Set<CompilationEnvironment> environments = new HashSet<>();
    private final ReadWriteLock lock = new ReentrantReadWriteLock();

    /* Text Document Service */

    @Override
    public void didOpen(DidOpenTextDocumentParams params) {
        CompletableFuture.runAsync(() -> {
            lockForWrite();
            try {
                Path path = PathUtil.toPath(params.getTextDocument().getUri());
                LogMessages.start("didOpen", path, logger);
                if (getEnv(path).isEmpty()) {
                    createEnv(path);
                }
                LogMessages.finish("didOpen", path, logger);
            } finally {
                unlockForWrite();
            }
        }).exceptionally(e -> {
            LogMessages.error("didOpen", params, e, logger);
            return null;
        });
    }

    @Override
    public void didChange(DidChangeTextDocumentParams params) {
        CompletableFuture.runAsync(() -> {
            lockForWrite();
            try {
                Path path = PathUtil.toPath(params.getTextDocument().getUri());
                LogMessages.start("didChange", path, logger);
                CompilationUnit unit = getUnit(path).orElseThrow();
                String source = params.getContentChanges().get(0).getText();
                Compilations.load(unit, source);
                LogMessages.finish("didChange", path, logger);
            } finally {
                unlockForWrite();
            }
        }).exceptionally(e -> {
            LogMessages.error("didChange", params, e, logger);
            return null;
        });
    }

    @Override
    public void didClose(DidCloseTextDocumentParams params) {
    }

    @Override
    public void didSave(DidSaveTextDocumentParams params) {
    }

    @Override
    public CompletableFuture<Either<List<CompletionItem>, CompletionList>> completion(CompletionParams params) {
        return CompletableFuture.supplyAsync(() -> {
            lockForRead();
            try {
                Path path = PathUtil.toPath(params.getTextDocument().getUri());
                LogMessages.start("completion", path, params.getPosition(), logger);
                CompilationUnit unit = getUnit(path).orElseThrow();
                CompletionList result = CompletionProvider.completion(unit, params);
                LogMessages.finish("completion", path, params.getPosition(), logger);
                return Either.<List<CompletionItem>, CompletionList>forRight(result);
            } finally {
                unlockForRead();
            }
        }).exceptionally(e -> {
            LogMessages.error("completion", params, e, logger);
            return null;
        });
    }

    @Override
    public CompletableFuture<CompletionItem> resolveCompletionItem(CompletionItem unresolved) {
        return CompletableFuture.supplyAsync(() -> {
            LogMessages.start("resolveCompletionItem", logger);
            // do nothing
            LogMessages.finish("resolveCompletionItem", logger);
            return unresolved;
        }).exceptionally(e -> {
            LogMessages.error("resolveCompletionItem", unresolved, e, logger);
            return null;
        });
    }

    @Override
    public CompletableFuture<Hover> hover(HoverParams params) {
        return CompletableFuture.supplyAsync(() -> {
            lockForRead();
            try {
                Path path = PathUtil.toPath(params.getTextDocument().getUri());
                LogMessages.start("hover", path, params.getPosition(), logger);
                CompilationUnit unit = getUnit(path).orElseThrow();
                Hover result = HoverProvider.hover(unit, params);
                LogMessages.finish("hover", path, params.getPosition(), logger);
                return result;
            } finally {
                unlockForRead();
            }
        }).exceptionally(e -> {
            LogMessages.error("hover", params, e, logger);
            return null;
        });
    }

    @Override
    public CompletableFuture<Either<List<? extends Location>, List<? extends LocationLink>>> definition(DefinitionParams params) {
        return CompletableFuture.supplyAsync(() -> {
            lockForRead();
            try {
                Path path = PathUtil.toPath(params.getTextDocument().getUri());
                LogMessages.start("definition", path, params.getPosition(), logger);
                CompilationUnit unit = getUnit(path).orElseThrow();
                List<LocationLink> result = DefinitionProvider.definition(unit, params);
                LogMessages.finish("definition", path, params.getPosition(), logger);
                return Either.<List<? extends Location>, List<? extends LocationLink>>forRight(result);
            } finally {
                unlockForRead();
            }
        }).exceptionally(e -> {
            LogMessages.error("definition", params, e, logger);
            return null;
        });
    }

    @Override
    public CompletableFuture<List<? extends Location>> references(ReferenceParams params) {
        return CompletableFuture.supplyAsync(() -> {
            lockForRead();
            try {
                Path path = PathUtil.toPath(params.getTextDocument().getUri());
                LogMessages.start("references", path, params.getPosition(), logger);
                CompilationUnit unit = getUnit(path).orElseThrow();
                List<Location> result = ReferencesProvider.references(unit, params);
                LogMessages.finish("references", path, params.getPosition(), logger);
                return result;
            } finally {
                unlockForRead();
            }
        }).handle((result, throwable) -> {
            if (throwable != null) {
                LogMessages.error("references", params, throwable, logger);
            }
            return result;
        });
    }

    @Override
    public CompletableFuture<List<Either<SymbolInformation, DocumentSymbol>>> documentSymbol(DocumentSymbolParams params) {
        return CompletableFuture.supplyAsync(() -> {
            lockForRead();
            try {
                Path path = PathUtil.toPath(params.getTextDocument().getUri());
                LogMessages.start("documentSymbol", path, logger);
                CompilationUnit unit = getUnit(path).orElseThrow();
                List<DocumentSymbol> result = DocumentSymbolProvider.documentSymbol(unit, params);
                LogMessages.finish("documentSymbol", path, logger);
                return result.stream()
                        .map(Either::<SymbolInformation, DocumentSymbol>forRight)
                        .toList();
            } finally {
                unlockForRead();
            }
        }).exceptionally(e -> {
            LogMessages.error("documentSymbol", params, e, logger);
            return null;
        });
    }

    @Override
    public CompletableFuture<SemanticTokens> semanticTokensFull(SemanticTokensParams params) {
        return CompletableFuture.supplyAsync(() -> {
            lockForRead();
            try {
                Path path = PathUtil.toPath(params.getTextDocument().getUri());
                LogMessages.start("semanticTokensFull", path, logger);
                CompilationUnit unit = getUnit(path).orElseThrow();
                SemanticTokens result = SemanticTokensProvider.semanticTokensFull(unit, params);
                LogMessages.finish("semanticTokensFull", path, logger);
                return result;
            } finally {
                unlockForRead();
            }
        }).exceptionally(e -> {
            LogMessages.error("semanticTokensFull", params, e, logger);
            return null;
        });
    }

    /* End Text Document Service */

    /* Workspace Service */

    @Override
    public void didChangeConfiguration(DidChangeConfigurationParams params) {
    }

    @Override
    public void didChangeWatchedFiles(DidChangeWatchedFilesParams params) {
        CompletableFuture.runAsync(() -> {
            lockForWrite();
            try {
                LogMessages.start("didChangeWatchedFiles", logger);
                for (FileEvent event : params.getChanges()) {
                    Path path = PathUtil.toPath(event.getUri());
                    getEnv(path).ifPresent(env -> {
                        switch (event.getType()) {
                            case Created -> {
                                CompilationUnit unit = env.createUnit(path);
                                Compilations.load(unit);
                            }
                            case Changed -> {
                                CompilationUnit unit = env.getUnit(path);
                                Compilations.load(unit);
                            }
                            case Deleted -> env.removeUnit(path);
                        }
                    });
                }
                LogMessages.finish("didChangeWatchedFiles", logger);
            } catch (Exception e) {
                LogMessages.error("didChangeWatchedFiles", e, logger);
            } finally {
                unlockForWrite();
            }
        });
    }

    @Override
    public void didChangeWorkspaceFolders(DidChangeWorkspaceFoldersParams params) {
    }

    /* End Workspace Service */

    private void createEnv(Path documentPath) {
        Path compilationRoot = PathUtil.findUpwardsOrSelf(documentPath, CompilationEnvironment.DEFAULT_ROOT_DIRECTORY);
        CompilationEnvironment env = new CompilationEnvironment(compilationRoot);
        Compilations.load(env);
        environments.add(env);
        checkDzs(env);
    }

    private Optional<CompilationEnvironment> getEnv(Path documentPath) {
        return environments.stream()
                .filter(env -> env.containsUnit(documentPath))
                .findFirst();
    }

    private Optional<CompilationUnit> getUnit(Path documentPath) {
        return getEnv(documentPath).map(env -> env.getUnit(documentPath));
    }

    private void checkDzs(CompilationEnvironment env) {
        if (Files.exists(env.getGeneratedRoot())) {
            logger.info("Cannot find .dzs file directory of environment: {}", env);
            LogMessages.info(L10N.getString("dzs_not_found"), ZenLanguageServer.getClient());
        }
    }

    private void lockForRead() {
        lock.readLock().lock();
    }

    private void unlockForRead() {
        lock.readLock().unlock();
    }

    private void lockForWrite() {
        lock.writeLock().lock();
    }

    private void unlockForWrite() {
        lock.writeLock().unlock();
    }

}

File: zenscript-language-server/src/main/kotlin/raylras/zen/lsp/ZenLanguageService.kt
package raylras.zen.lsp

import org.eclipse.lsp4j.*
import org.eclipse.lsp4j.jsonrpc.messages.Either
import org.eclipse.lsp4j.services.TextDocumentService
import org.eclipse.lsp4j.services.WorkspaceService
import org.slf4j.LoggerFactory
import raylras.zen.lsp.StandardIOLauncher.CLIENT
import raylras.zen.lsp.provider.*
import raylras.zen.model.*
import raylras.zen.util.*
import raylras.zen.util.l10n.L10N
import java.nio.file.Path
import java.util.concurrent.CompletableFuture
import java.util.concurrent.locks.ReentrantReadWriteLock
import kotlin.io.path.notExists
import kotlin.time.measureTime
import kotlin.time.measureTimedValue

class ZenLanguageService : TextDocumentService, WorkspaceService {
    private val logger = LoggerFactory.getLogger(javaClass)
    private val environments = HashSet<CompilationEnvironment>()
    private val lock = ReentrantReadWriteLock()

    //region Text Document Service
    override fun didOpen(params: DidOpenTextDocumentParams) {
        val uri = params.textDocument.uri
        logger.start(::didOpen, uri)
        lockForWrite()
        try {
            measureTime {
                val path = uri.toPath()
                getEnv(path) ?: createEnv(path)
            }.let {
                logger.finish(::didOpen, uri, duration = it)
            }
        } catch (e: Exception) {
            logger.fail(::didOpen, uri, throwable = e)
        } finally {
            unlockForWrite()
        }
    }

    override fun didChange(params: DidChangeTextDocumentParams) {
        val uri = params.textDocument.uri
        logger.start(::didChange, uri)
        lockForWrite()
        try {
            measureTime {
                val unit = uri.toCompilationUnit()!!
                val source = params.contentChanges[0].text
                unit.load(source)
            }.let {
                logger.finish(::didChange, uri, duration = it)
            }
        } catch (e: Exception) {
            logger.fail(::didChange, uri, throwable = e)
        } finally {
            unlockForWrite()
        }
    }

    override fun didClose(params: DidCloseTextDocumentParams) {
    }

    override fun didSave(params: DidSaveTextDocumentParams) {
    }

    override fun completion(params: CompletionParams): CompletableFuture<Either<List<CompletionItem>, CompletionList>> {
        return CompletableFuture.supplyAsync {
            val uri = params.textDocument.uri
            val pos = params.position
            logger.start(::completion, uri, pos)
            lockForRead()
            try {
                measureTimedValue {
                    val unit = uri.toCompilationUnit()!!
                    CompletionProvider.completion(unit, params)
                }.let { (value, duration) ->
                    logger.finish(::completion, uri, pos, duration)
                    Either.forRight(value)
                }
            } catch (e: Exception) {
                logger.fail(::completion, uri, pos, e)
                null
            } finally {
                unlockForRead()
            }
        }
    }

    override fun resolveCompletionItem(unresolved: CompletionItem): CompletableFuture<CompletionItem> {
        return CompletableFuture.supplyAsync {
            logger.start(::resolveCompletionItem)
            try {
                measureTimedValue {
                    CompletionProvider.resolveCompletionItem(unresolved)
                }.let { (value, duration) ->
                    logger.finish(::resolveCompletionItem, duration = duration)
                    value
                }
            } catch (e: Exception) {
                logger.fail(::resolveCompletionItem, throwable = e)
                unresolved
            }
        }
    }

    override fun hover(params: HoverParams): CompletableFuture<Hover> {
        return CompletableFuture.supplyAsync {
            val uri = params.textDocument.uri
            val pos = params.position
            logger.start(::hover, uri, pos)
            lockForRead()
            try {
                measureTimedValue {
                    val unit = uri.toCompilationUnit()!!
                    HoverProvider.hover(unit, params)
                }.let { (value, duration) ->
                    logger.finish(::hover, uri, pos, duration)
                    value
                }
            } catch (e: Exception) {
                logger.fail(::hover, uri, pos, e)
                null
            } finally {
                unlockForRead()
            }
        }
    }

    override fun definition(params: DefinitionParams): CompletableFuture<Either<List<Location>, List<LocationLink>>> {
        return CompletableFuture.supplyAsync {
            val uri = params.textDocument.uri
            val pos = params.position
            logger.start(::definition, uri, pos)
            lockForRead()
            try {
                measureTimedValue {
                    val unit = uri.toCompilationUnit()!!
                    DefinitionProvider.definition(unit, params)
                }.let { (value, duration) ->
                    logger.finish(::definition, uri, pos, duration)
                    Either.forRight(value)
                }
            } catch (e: Exception) {
                logger.fail(::definition, uri, pos, e)
                null
            } finally {
                unlockForRead()
            }
        }
    }

    override fun references(params: ReferenceParams): CompletableFuture<List<Location>> {
        return CompletableFuture.supplyAsync {
            val uri = params.textDocument.uri
            val pos = params.position
            logger.start(::references, uri, pos)
            lockForRead()
            try {
                measureTimedValue {
                    val unit = uri.toCompilationUnit()!!
                    ReferencesProvider.references(unit, params)
                }.let { (value, duration) ->
                    logger.finish(::references, uri, pos, duration)
                    value
                }
            } catch (e: Exception) {
                logger.fail(::references, uri, pos, e)
                null
            } finally {
                unlockForRead()
            }
        }
    }

    override fun documentSymbol(params: DocumentSymbolParams): CompletableFuture<List<Either<SymbolInformation, DocumentSymbol>>> {
        return CompletableFuture.supplyAsync {
            val uri = params.textDocument.uri
            logger.start(::documentSymbol, uri)
            lockForRead()
            try {
                measureTimedValue {
                    val unit = uri.toCompilationUnit()!!
                    DocumentSymbolProvider.documentSymbol(unit, params)
                }.let { (value, duration) ->
                    logger.finish(::documentSymbol, uri, duration = duration)
                    value.map { Either.forRight(it) }
                }
            } catch (e: Exception) {
                logger.fail(::documentSymbol, uri, throwable = e)
                null
            } finally {
                unlockForRead()
            }
        }
    }

    override fun semanticTokensFull(params: SemanticTokensParams): CompletableFuture<SemanticTokens> {
        return CompletableFuture.supplyAsync {
            val uri = params.textDocument.uri
            logger.start(::semanticTokensFull, uri)
            lockForRead()
            try {
                measureTimedValue {
                    val unit = uri.toCompilationUnit()!!
                    SemanticTokensProvider.semanticTokensFull(unit, params)
                }.let { (value, duration) ->
                    logger.finish(::semanticTokensFull, uri, duration = duration)
                    value
                }
            } catch (e: Exception) {
                logger.fail(::semanticTokensFull, uri, throwable = e)
                null
            } finally {
                unlockForRead()
            }
        }
    }

    override fun semanticTokensRange(params: SemanticTokensRangeParams): CompletableFuture<SemanticTokens> {
        return CompletableFuture.supplyAsync {
            val uri = params.textDocument.uri
            logger.start(::semanticTokensRange, uri)
            lockForRead()
            try {
                measureTimedValue {
                    val unit = uri.toCompilationUnit()!!
                    SemanticTokensProvider.semanticTokensRange(unit, params)
                }.let { (value, duration) ->
                    logger.finish(::semanticTokensRange, uri, duration = duration)
                    value
                }
            } catch (e: Exception) {
                logger.fail(::semanticTokensRange, uri, throwable = e)
                null
            } finally {
                unlockForRead()
            }
        }
    }

    override fun inlayHint(params: InlayHintParams): CompletableFuture<List<InlayHint>> {
        return CompletableFuture.supplyAsync {
            val uri = params.textDocument.uri
            logger.start(::inlayHint, uri)
            lockForRead()
            try {
                measureTimedValue {
                    val unit = uri.toCompilationUnit()!!
                    InlayHintProvider.inlayHint(unit, params)
                }.let { (value, duration) ->
                    logger.finish(::inlayHint, uri, duration = duration)
                    value
                }
            } catch (e: Exception) {
                logger.fail(::inlayHint, uri, throwable = e)
                null
            } finally {
                unlockForRead()
            }
        }
    }
    //endregion

    //region Workspace Service
    override fun didChangeConfiguration(params: DidChangeConfigurationParams) {
    }

    override fun didChangeWatchedFiles(params: DidChangeWatchedFilesParams) {
        logger.start(::didChangeWatchedFiles)
        lockForWrite()
        try {
            measureTime {
                for (event in params.changes) {
                    val path = event.uri.toPath()
                    getEnv(path)?.let { env ->
                        when (event.type) {
                            FileChangeType.Created -> {
                                createUnit(path, env).load()
                            }

                            FileChangeType.Changed -> {
                                env.unitMap[path]!!.load()
                            }

                            FileChangeType.Deleted -> env.unitMap.remove(path)

                            else -> {}
                        }
                    }
                }
            }.let {
                logger.finish(::didChangeWatchedFiles, duration = it)
            }
        } catch (e: Exception) {
            logger.fail(::didChangeWatchedFiles, throwable = e)
        } finally {
            unlockForWrite()
        }
    }

    override fun didChangeWorkspaceFolders(params: DidChangeWorkspaceFoldersParams) {
    }
    //endregion

    //region Private
    private fun createEnv(documentPath: Path) {
        val compilationRoot = findUpwardsOrSelf(documentPath, DEFAULT_ROOT_DIRECTORY)
        val env = CompilationEnvironment(compilationRoot)
        env.load()
        environments.add(env)
        checkDzs(env)
    }

    private fun getEnv(documentPath: Path): CompilationEnvironment? {
        return environments.find { it.unitMap.containsKey(documentPath) }
    }

    private fun getUnit(documentPath: Path): CompilationUnit? {
        return getEnv(documentPath)?.let { it.unitMap[documentPath] }
    }

    private fun checkDzs(env: CompilationEnvironment) {
        if (env.generatedRoot.notExists()) {
            logger.info("Cannot find .dzs file directory of environment: {}", env)
            CLIENT.info(L10N.localize("dzs_not_found"))
        }
    }

    private fun lockForRead() {
        lock.readLock().lock()
    }

    private fun unlockForRead() {
        lock.readLock().unlock()
    }

    private fun lockForWrite() {
        lock.writeLock().lock()
    }

    private fun unlockForWrite() {
        lock.writeLock().unlock()
    }

    private fun String.toCompilationUnit(): CompilationUnit? {
        return getUnit(this.toPath())
    }

    private fun String.toPath(): Path {
        return toPath(this)
    }
    //endregion
}
